fairmq/nanomsg/FairMQSocketNN.h

```diff
@@ -71,6 +89,8 @@ class FairMQSocketNN : public FairMQSocket
     std::atomic<unsigned long> fMessagesRx;
 
     static std::atomic<bool> fInterrupted;
+    static bool fHasLingered;
+    static int fLingerDuration;
```

Comment:
Tracking the linger with global state seems dirty. My counter proposal would be to put `fLingerDuration` as a non-static member and move the actual linger implementation to the `FairMQTransportFactoryNN`. For this, it needs to track the created sockets. On `RESET_DEVICE` we could e.g. call `FairMQTransportFactory::Reset` on all transports, which - in the case of nanomsg - calculates the max of all linger values on all nn sockets and sleeps that long. What do you think?

Reply:
That leads to another discussion: Should the transport factory give up ownership on the socket. In case of zeromq at least this could break, if one destructs the transport factory before the socket.

Reply:
> If we do go for owning factories, I would go with unique_ptrs for sockets inside the factory and return non-owning raw ptrs. When device is present we control the lifetime of both factories and sockets, so no additional checks would be necessary at least inside of our classes.

I agree, would do it the same.

Reply:
Is it ok, if this blocks the other commits in this PR? If you want, you could open a seperate PR just with this commit, so the other ones can be merged already.

Reply:
Merging without this will cause the nanomsg test to fail a lot. So I will change the rateLogger to use regular sleep again, until nanomsg is fixed.

Reply:
Done, this is now only WaitFor and some changes in the tests.
Nanomsg linger changes will follow in separate PR.